File: model_definitions/f5c6d40d-ecba-4820-b2c3-85af85cab552/model_modules/training.R
LoadPackages <- function() {
    library("h2o")
    library("DBI")
    library("dplyr")
    library("tdplyr")
}

suppressPackageStartupMessages(LoadPackages())

Connect2Vantage <- function() {
    # Create Vantage connection using tdplyr
    con <- td_create_context(host = Sys.getenv("AOA_CONN_HOST"),
                             uid = Sys.getenv("AOA_CONN_USERNAME"),
                             pwd = Sys.getenv("AOA_CONN_PASSWORD"),
                             dType = 'native'
    )

    # Set connection context
    td_set_context(con)

    con
}

train <- function(data_conf, model_conf, ...) {
    print("Training model...")

    # Connect to Vantage
    con <- Connect2Vantage()

    # Create tibble from table in Vantage
    if ("schema" %in% data_conf) {
        table_name <- in_schema(data_conf$schema, data_conf$table)
    } else {
        table_name <- data_conf$table
    }
    table <- tbl(con, table_name)
    
    predictors <- c('age', 'job', 'marital', 'education', 'default', 'balance', 'housing', 'loan', 'y')
    target <- 'y'

    # Create dataframe from tibble, selecting the necessary columns and mutating as necessary
    print("Load dataset")
    data <- table %>% select(all_of(predictors)) %>% as.data.frame()
    data$age <- as.integer(data$age)
    data$job <- as.factor(data$job)
    data$marital <- as.factor(data$marital)
    data$education <- as.factor(data$education)
    data$default <- as.factor(data$default)
    data$balance <- as.integer(data$balance)
    data$housing <- as.factor(data$housing)
    data$loan <- as.factor(data$loan)
    data$y <- as.factor(data$y)
    data
    
    # Load hyperparameters from model configuration
    hyperparams <- model_conf[["hyperParameters"]]

    # Initialize and convert dataframe to h2o
    print("Init h2o")
    h2o.init(nthreads = -1)
    train.hex <- as.h2o(data)
    splits <- h2o.splitFrame(train.hex, 0.75, seed=1234)

    # Train model
    model <- h2o.gbm(
        x = predictors,
        y = target, 
        training_frame = splits[[1]],
        categorical_encoding = 'auto', 
        ntrees = hyperparams$ntrees, 
        max_depth = hyperparams$max_depth, 
        min_rows = hyperparams$min_rows, 
        learn_rate = hyperparams$learn_rate
    )
    print("Model Trained!")

    # Save trained model
    print("Saving trained model...")
    #saveRDS(model, "artifacts/output/model.rds")
    
    # Save trained model in h2o format
    output.dir <- getwd()
    path.value <- file.path(output.dir, "artifacts/output")
    h2o.saveModel(object = model, path = path.value, force = TRUE)
    name <- file.path(path.value, "model.h2o") # destination file name at the same folder location
    file.rename(file.path(path.value, model@model_id), name)
}

File: model_definitions/eb0fffd2-1cbc-466a-9fe9-e490bd183751/model_modules/training.py
from teradataml import create_context
from teradataml.dataframe.dataframe import DataFrame
from aoa.stats import stats
from aoa.util.artefacts import save_plot
import os
import h2o
from h2o.estimators.gbm import H2OGradientBoostingEstimator


def train(data_conf, model_conf, **kwargs):
    hyperparams = model_conf["hyperParameters"]

    create_context(host=os.environ["AOA_CONN_HOST"],
                   username=os.environ["AOA_CONN_USERNAME"],
                   password=os.environ["AOA_CONN_PASSWORD"],
                   database=data_conf["schema"] if "schema" in data_conf and data_conf["schema"] != "" else None)

    feature_names = ['age', 'job', 'marital', 'education', 'default', 'balance', 'housing', 'loan']
    target_name = 'y'
    categorical_names = ['age', 'job', 'marital', 'education', 'default', 'housing', 'loan', 'y']

    # read training dataset from Teradata and convert to pandas
    h2o.init()
    train_df = DataFrame(data_conf["table"])
    train_df = train_df.select([feature_names + [target_name]])
    train_hdf = h2o.H2OFrame(train_df.to_pandas())

    # split data into X and y and factorize
    X_train, y_train = train_hdf.split_frame(ratios=[.7])
    X_train[target_name] = X_train[target_name].asfactor()
    y_train[target_name] = y_train[target_name].asfactor()

    print("Starting training...")

    # Execute AutoML on training data
    model = H2OGradientBoostingEstimator(categorical_encoding = 'auto', ntrees = hyperparams['ntrees'], max_depth = hyperparams['max_depth'], min_rows = hyperparams['min_rows'], learn_rate = hyperparams['learn_rate'])
    model.train(x=feature_names, y=target_name, training_frame=X_train)

    print("Finished training")

    # export model artefacts
    current_path = os.path.abspath(os.getcwd())
    artifacts_path = "artifacts/output"
    old_model = h2o.save_model(model, artifacts_path)
    new_model = os.path.join(current_path, artifacts_path, "model.h2o")
    if os.path.isfile(new_model):
        print("The file already exists")
    else:
        # Rename the file
        os.rename(old_model, new_model)
        
    # Saving as h2o mojo
    mojo = model.download_mojo(path=artifacts_path, get_genmodel_jar=True)
    new_mojo = os.path.join(current_path, artifacts_path, "mojo.zip")
    if os.path.isfile(new_mojo):
        print("The file already exists")
    else:
        # Rename the file
        os.rename(mojo, new_mojo)

    # Convert mojo to pmml
    cmd = "wget https://aoa-public-files.s3.amazonaws.com/jpmml-h2o-executable-1.1-SNAPSHOT.jar && java -jar ./jpmml-h2o-executable-1.1-SNAPSHOT.jar --mojo-input {} --pmml-output {}".format(new_mojo, os.path.join(current_path, artifacts_path, "model.pmml"))
    result = os.system(cmd)
    if result > 0:
        raise OSError(result, "Error while trying to convert mojo to pmml")
    
    print("Saved trained model")

    model.varimp_plot()
    save_plot('feature_importance.png')
    fi = model.varimp(True)
    fix = fi[['variable','scaled_importance']]
    fis = fix.to_dict('records')
    feature_importance = {v['variable']:v['scaled_importance'] for (k,v) in enumerate(fis)}

    stats.record_training_stats(train_df,
                       features=feature_names,
                       predictors=[target_name],
                       categorical=categorical_names,
                       importance=feature_importance)
